xcdat/temporal_avg.py

```diff
@@ -0,0 +1,988 @@
+        and typical weather for a particular frequency.
+
+        Parameters
+        ----------
+        data_var: Optional[str], optional
+            The key of the data variable in the dataset to calculate departures
+            for. If None, xcdat attempts to infer the desired data variable
+            using the Dataset "xcdat_infer" attr (if defined), by default None.
+
+        Returns
+        -------
+        xr.Dataset
+            The Dataset containing the departures for a data var's climatology.
+
+        Examples
+        --------
+        Import temporal averaging functionality:
+
+        >>> import xcdat
+
+        Get departure for an annual climatology:
+
+        >>> ds_climo = ds.temporal.temporal_avg("climatology", "month", "ts")
+        >>> ds_departure = ds_climo.temporal.departure("ts")
+
+        Get data variable's temporal averaging operation attributes:
+
+        >>> ds_season_with_djf.ts.attrs
+        {
+            'operation': 'temporal_avg',
+            'mode': 'departure',
+            'frequency': 'season',
+            'groupby': 'season',
+            'weighted': 'True',
+            'centered_time': 'True',
+            'dec_mode': 'DJF',
+            'drop_incomplete_djf': 'True'
+        }
+        """
+        ds = self._dataset.copy()
+        dv_climo = get_data_var(ds, data_var)
+
+        # Reuse the object attributes using the data variable's climatology
+        # operation attributes
+        attrs = dv_climo.attrs
+        operation = attrs.get("operation")
+        if operation is None:
+            raise KeyError(
+                f"The data var, '{dv_climo.name}', does not contain the "
+                "'operation' attribute for temporal averaging was performed. "
+                f"Make sure to run the climatology on '{dv_climo.name}' first "
+                "before calculating its departure."
+            )
+        if attrs["mode"] != "climatology":
+            raise ValueError(
+                "Can only calculate departures for a data variable's climatology."
+            )
+        self._set_obj_attrs(
+            "departure",
+            attrs["freq"],
+            str_to_bool(attrs["weighted"]),
+            str_to_bool(attrs["centered_time"]),
+            {
+                "dec_mode": attrs.get("dec_mode", "DJF"),
+                "drop_incomplete_djf": str_to_bool(
+                    attrs.get("drop_incomplete_djf", "True")
+                ),
+                "custom_seasons": attrs.get("custom_seasons", None),
+            },
+        )
+        self._time_bounds: xr.DataArray = ds.time_bnds.copy()
+
+        # Get departures by calculating the difference between observations and
+        # the climatology.
+        departures = ds.copy()
+        dv_og = ds[f"{dv_climo.name}_original"].copy().rename("ts")
+        dv_obs = self._group_observations(dv_og)
+
+        with xr.set_options(keep_attrs=True):
+            dv_departures = dv_obs - dv_climo
```

Comment:
I think something is going awry in how I am attempting to calculate seasonal departures:

```
fn = '/p/user_pub/work/E3SM/1_0/historical/1deg_atm_60-30km_ocean/atmos/180x360/time-series/mon/ens1/v5/TS_185001_201412.nc'
mode = 'timeseries'
freq = 'season'
weighted = True
centered_time = True
# custom_seasons = [["Jan", "Feb", "Mar"], ["Apr", "May", "Jun"], ["Jul", "Aug", "Sep"], ["Oct", "Nov", "Dec"]]
season_config = {"dec_mode": "DJF", "drop_incomplete_djf": True,
                 "custom_seasons": None}
ds = xcdat.open_dataset(fn, data_var='TS')
ds = ds.xcdat.spatial_avg()
ds_climo = ds.temporal.temporal_avg("climatology", "season")
ds_departure = ds_climo.temporal.departure()
ds.close()
```

The output is not what I expect. I expect it to have roughly size `len(ds.TS)/4`, but I am getting the full time series length (1980 months). 

One other thing that is no problem, but I didn't realize at first is that the seasonal output starts with MAM. I expected it to start with DJF. I think this is fine, but noting it in case it is a surprise to others.

Reply:
Okay – this appears to remove the seasonal anomaly from each month, but an additional step is required to get the seasonal anomaly:

```
ds_departure_seasonal = ds_departure.temporal.temporal_avg("timeseries", "season")
```

Reply:
Seasonal output starting with DJF perhaps is more intuitive. I would propose to change the order ..

Reply:
> Okay – this appears to remove the seasonal anomaly from each month, but an additional step is required to get the seasonal anomaly:
> 
> ```
> ds_departure_seasonal = ds_departure.temporal.temporal_avg("timeseries", "season")
> ```

As mentioned on Slack, this can be addressed by adding this call to the end of the `departure()` method with the same `freq`.

> One other thing that is no problem, but I didn't realize at first is that the seasonal output starts with MAM. I expected it to start with DJF. I think this is fine, but noting it in case it is a surprise to others.

> Seasonal output starting with DJF perhaps is more intuitive. I would propose to change the order .. Actually, I think it is fine to start with MAM. NVM...

The seasonal output starts at MAM because when the leading and trailing incomplete DJF seasons are removed. The month for the first timestamp is March, so the grouping starts at MAM.

A one-liner call to xarray's sort method can be implemented.